# Test for Fixed Effects

## Load Libraries and Dataset

### Libraries

In [1]:
import numpy as np
import pandas as pd
import scipy.stats as stats
import statsmodels.api as sm
from linearmodels.panel import PanelOLS
from linearmodels.panel import PooledOLS

### Dataset and PanelOLS Object

In [2]:
sgx = pd.read_csv("data/clean_sgx.csv", index_col= None)
sgx = sgx[['Company Code', 'Year', 'LEVERAGE',
           'SIZE', 'PROFITABILITY', 'TANG', 'LIQUID', 'MCAP', 'SOLV']]
sgx['year'] = pd.Categorical(sgx['Year'])
sgx['compcode'] = pd.Categorical(sgx['Company Code'])
sgx = sgx.set_index(['Company Code', 'Year'])

endo = sgx.LEVERAGE
exog_vars = ['SIZE', 'PROFITABILITY', 'TANG', 'LIQUID', 'MCAP', 'SOLV']
exog = sm.add_constant(sgx[exog_vars])

pooled_mod = PooledOLS(endo, exog)
pooled_res = pooled_mod.fit()

In [13]:
sgx.head(30)


LEVERAGE       SIZE  PROFITABILITY      TANG    LIQUID  \
Company Code Year                                                             
3            2015.0 -0.235443  17.927564      -0.094659  0.176726  1.563892   
             2016.0 -0.206973  17.984885      -0.010839  0.096761  1.505202   
             2017.0 -0.238219  18.136273      -0.013888  0.081307  1.343012   
             2018.0 -0.259016  18.114074       0.013371  0.079477  1.351929   
             2019.0 -0.222870  17.978821       0.078838  0.113450  1.558678   
             2020.0 -0.353373  18.184655       0.126747  0.077289  1.653512   
             2021.0 -0.354726  18.488321       0.118694  0.060646  1.582983   
             2022.0 -0.473794  18.382137       0.079766  0.063917  1.825263   
10           2015.0 -0.027551  17.648089       0.029971  0.094432  1.726605   
             2016.0  0.026441  17.779968       0.035823  0.151546  1.365213   
             2017.0  0.071474  17.879947       0.042540  0.163954  1.398815   
             2018.0  0.124074  17.825039      -0.001821  0.153358  1.303544   
             2019.0  0.267393  17.789708      -0.071748  0.298266  1.079697   
             2020.0  0.300076  17.569981       0.006588  0.283695  1.066758   
             2021.0  0.231850  17.314345      -0.032096  0.229591  1.047314   
             2022.0  0.234610  17.319097       0.068364  0.187101  1.020956   
12           2015.0  0.199280  16.916278      -0.037227  0.042269  0.536030   
             2016.0  0.073882  16.805376      -0.034099  0.039883  0.922022   
             2017.0  0.110147  16.475071      -0.230511  0.041917  1.021277   
             2018.0  0.215825  16.442061      -0.150514  0.023723  0.591639   
             2019.0  0.282343  16.787616       0.173040  0.056827  0.953543   
             2020.0  0.334869  16.820698      -0.154581  0.040020  0.921919   
             2021.0  0.336646  16.516268       0.065744  0.031697  0.539579   
             2022.0  0.256070  16.609800      -0.067886  0.033637  1.397857   
14           2015.0  0.226407  16.374310      -0.290508  0.111944  1.161707   
             2016.0  0.131923  16.313605      -0.108950  0.105254  1.003601   
             2017.0 -0.230777  16.558217      -0.167373  0.097032  1.170574   
             2018.0 -0.100940  16.666088      -0.105473  0.291681  0.815182   
             2019.0 -0.024855  16.548582      -0.177847  0.346863  0.558211   
             2020.0  0.019507  16.513552      -0.060864  0.394418  0.477831   

                             MCAP     SOLV    year compcode  
Company Code Year                                            
3            2015.0  3.726956e+07  -31.451  2015.0        3  
             2016.0  6.995037e+06   -7.719  2016.0        3  
             2017.0  7.919165e+06   -8.678  2017.0        3  
             2018.0  9.212378e+06    0.966  2018.0        3  
             2019.0  1.931658e+07   11.465  2019.0        3  
             2020.0  1.459661e+07   61.729  2020.0        3  
             2021.0  2.725603e+07  194.000  2021.0        3  
             2022.0  4.331235e+07   56.147  2022.0        3  
10           2015.0  1.888834e+07   -1.277  2015.0       10  
             2016.0  8.499886e+06    1.285  2016.0       10  
             2017.0  9.514661e+06    0.764  2017.0       10  
             2018.0  6.367345e+06   -2.850  2018.0       10  
             2019.0  3.209531e+06   -5.687  2019.0       10  
             2020.0  2.340626e+06   -3.112  2020.0       10  
             2021.0  1.254450e+06   -5.483  2021.0       10  
             2022.0  1.134966e+06   -0.086  2022.0       10  
12           2015.0  3.690988e+07   -2.039  2015.0       12  
             2016.0  4.643639e+07   -3.355  2016.0       12  
             2017.0  5.263948e+07  -24.447  2017.0       12  
             2018.0  3.172811e+07   -3.424  2018.0       12  
             2019.0  2.625733e+07    2.233  2019.0       12  
             2020.0  3.505423e+07   -2.289  2020.0       12  
    

## F-test for Fixed Effects (Entity-clustered Robust)

In [14]:
def fixedeffect(endo = endo, exog = exog,
                entity_eff = True, time_eff = True,
                cov_type = 'clustered', cluster_entity = True, **cov_kwargs):
    mod = PanelOLS(endo, exog,
                   entity_effects=entity_eff,
                   time_effects=time_eff)
    res = mod.fit(cov_type= cov_type, cluster_entity = cluster_entity, **cov_kwargs)

    return res

### Joint test for two-way FE

In [15]:
fixedeffect().summary

<class 'linearmodels.compat.statsmodels.Summary'>
"""
                          PanelOLS Estimation Summary                           
================================================================================
Dep. Variable:               LEVERAGE   R-squared:                        0.2294
Estimator:                   PanelOLS   R-squared (Between):              0.1387
No. Observations:                1728   R-squared (Within):               0.2397
Date:                Mon, Apr 01 2024   R-squared (Overall):              0.1569
Time:                        00:51:05   Log-likelihood                    1628.0
Cov. Estimator:             Clustered                                           
                                        F-statistic:                      74.386
Entities:                         216   P-value                           0.0000
Avg Obs:                       8.0000   Distribution:                  F(6,1499)
Min Obs:                       8.0000                                           
Max Obs:                       8.0000   F-statistic (robust):             15.283
                                        P-value                           0.0000
Time periods:                       8   Distribution:                  F(6,1499)
Avg Obs:                       216.00                                           
Min Obs:                       216.00                                           
Max Obs:                       216.00                                           
                                                                                
                               Parameter Estimates                               
=================================================================================
               Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
---------------------------------------------------------------------------------
const            -1.5191     0.5081    -2.9899     0.0028     -2.5157     -0.5225
SIZE              0.0816     0.0259     3.1492     0.0017      0.0308      0.1325
PROFITABILITY    -0.2288     0.0562    -4.0735     0.0000     -0.3390     -0.1186
TANG              0.4097     0.0849     4.8235     0.0000      0.2431      0.5763
LIQUID           -0.0228     0.0066    -3.4833     0.0005     -0.0357     -0.0100
MCAP          -3.265e-12  3.564e-12    -0.9161     0.3598  -1.026e-11   3.727e-12
SOLV             -0.0002     0.0001    -1.5528     0.1207     -0.0004   4.813e-05
=================================================================================

F-test for Poolability: 25.310
P-value: 0.0000
Distribution: F(222,1499)

Included effects: Entity, Time
"""

***Joint F-test for FE for eneity cluster on***

F-stat: **74.386**

F-stat (robust): **15.283**

### Conditional test for entity FE

In [16]:
fixedeffect(entity_eff=True, time_eff=False).summary

<class 'linearmodels.compat.statsmodels.Summary'>
"""
                          PanelOLS Estimation Summary                           
================================================================================
Dep. Variable:               LEVERAGE   R-squared:                        0.2401
Estimator:                   PanelOLS   R-squared (Between):              0.1308
No. Observations:                1728   R-squared (Within):               0.2401
Date:                Mon, Apr 01 2024   R-squared (Overall):              0.1504
Time:                        00:51:24   Log-likelihood                    1615.1
Cov. Estimator:             Clustered                                           
                                        F-statistic:                      79.286
Entities:                         216   P-value                           0.0000
Avg Obs:                       8.0000   Distribution:                  F(6,1506)
Min Obs:                       8.0000                                           
Max Obs:                       8.0000   F-statistic (robust):             18.662
                                        P-value                           0.0000
Time periods:                       8   Distribution:                  F(6,1506)
Avg Obs:                       216.00                                           
Min Obs:                       216.00                                           
Max Obs:                       216.00                                           
                                                                                
                               Parameter Estimates                               
=================================================================================
               Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
---------------------------------------------------------------------------------
const            -1.5595     0.4675    -3.3358     0.0009     -2.4765     -0.6425
SIZE              0.0834     0.0238     3.5041     0.0005      0.0367      0.1301
PROFITABILITY    -0.2270     0.0543    -4.1783     0.0000     -0.3336     -0.1205
TANG              0.4346     0.0807     5.3864     0.0000      0.2763      0.5928
LIQUID           -0.0228     0.0062    -3.6503     0.0003     -0.0350     -0.0105
MCAP          -4.195e-12  3.294e-12    -1.2735     0.2030  -1.066e-11   2.266e-12
SOLV             -0.0002     0.0001    -1.7470     0.0808     -0.0004   2.405e-05
=================================================================================

F-test for Poolability: 25.763
P-value: 0.0000
Distribution: F(215,1506)

Included effects: Entity
"""

***Conditional F-test for entity-FE for eneity cluster on***

F-stat: **79.286**

F-stat (robust): **18.662**

### Conditional test for time FE

In [17]:
fixedeffect(entity_eff=False, time_eff=True).summary

<class 'linearmodels.compat.statsmodels.Summary'>
"""
                          PanelOLS Estimation Summary                           
================================================================================
Dep. Variable:               LEVERAGE   R-squared:                        0.3616
Estimator:                   PanelOLS   R-squared (Between):              0.4295
No. Observations:                1728   R-squared (Within):               0.0537
Date:                Mon, Apr 01 2024   R-squared (Overall):              0.3621
Time:                        00:52:08   Log-likelihood                    285.77
Cov. Estimator:             Clustered                                           
                                        F-statistic:                      161.80
Entities:                         216   P-value                           0.0000
Avg Obs:                       8.0000   Distribution:                  F(6,1714)
Min Obs:                       8.0000                                           
Max Obs:                       8.0000   F-statistic (robust):             44.523
                                        P-value                           0.0000
Time periods:                       8   Distribution:                  F(6,1714)
Avg Obs:                       216.00                                           
Min Obs:                       216.00                                           
Max Obs:                       216.00                                           
                                                                                
                               Parameter Estimates                               
=================================================================================
               Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
---------------------------------------------------------------------------------
const            -0.5893     0.1405    -4.1945     0.0000     -0.8649     -0.3138
SIZE              0.0419     0.0068     6.1186     0.0000      0.0284      0.0553
PROFITABILITY    -0.4509     0.1061    -4.2491     0.0000     -0.6591     -0.2428
TANG              0.1898     0.0521     3.6417     0.0003      0.0876      0.2921
LIQUID           -0.0547     0.0076    -7.2022     0.0000     -0.0696     -0.0398
MCAP          -8.712e-12  4.074e-12    -2.1383     0.0326   -1.67e-11  -7.209e-13
SOLV             -0.0007     0.0001    -5.0611     0.0000     -0.0009     -0.0004
=================================================================================

F-test for Poolability: 1.0642
P-value: 0.3842
Distribution: F(7,1714)

Included effects: Time
"""

***Conditional F-test for time-FE for eneity cluster on***

F-stat: **161.80**

F-stat (robust): **44.523**

### Marginal test for entity FE

In [18]:
exog_vars_marginal_entity = ['SIZE', 'PROFITABILITY', 'TANG', 'LIQUID', 'MCAP', 'SOLV', 'year']
exog_marginal_entity = sm.add_constant(sgx[exog_vars_marginal_entity])
fixedeffect(exog = exog_marginal_entity, time_eff = False).summary

<class 'linearmodels.compat.statsmodels.Summary'>
"""
                          PanelOLS Estimation Summary                           
================================================================================
Dep. Variable:               LEVERAGE   R-squared:                        0.2513
Estimator:                   PanelOLS   R-squared (Between):              0.1387
No. Observations:                1728   R-squared (Within):               0.2513
Date:                Mon, Apr 01 2024   R-squared (Overall):              0.1589
Time:                        00:56:23   Log-likelihood                    1628.0
Cov. Estimator:             Clustered                                           
                                        F-statistic:                      38.706
Entities:                         216   P-value                           0.0000
Avg Obs:                       8.0000   Distribution:                 F(13,1499)
Min Obs:                       8.0000                                           
Max Obs:                       8.0000   F-statistic (robust):             13.731
                                        P-value                           0.0000
Time periods:                       8   Distribution:                 F(13,1499)
Avg Obs:                       216.00                                           
Min Obs:                       216.00                                           
Max Obs:                       216.00                                           
                                                                                
                               Parameter Estimates                               
=================================================================================
               Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
---------------------------------------------------------------------------------
const            -1.5240     0.4729    -3.2227     0.0013     -2.4516     -0.5964
SIZE              0.0816     0.0242     3.3675     0.0008      0.0341      0.1292
PROFITABILITY    -0.2288     0.0525    -4.3559     0.0000     -0.3318     -0.1258
TANG              0.4097     0.0794     5.1578     0.0000      0.2539      0.5655
LIQUID           -0.0228     0.0061    -3.7247     0.0002     -0.0348     -0.0108
MCAP          -3.265e-12  3.333e-12    -0.9796     0.3275  -9.804e-12   3.273e-12
SOLV             -0.0002     0.0001    -1.6604     0.0970     -0.0004   3.316e-05
year.2016.0       0.0015     0.0066     0.2350     0.8142     -0.0114      0.0145
year.2017.0      -0.0068     0.0101    -0.6774     0.4983     -0.0267      0.0130
year.2018.0       0.0031     0.0105     0.2983     0.7655     -0.0174      0.0237
year.2019.0       0.0305     0.0113     2.6993     0.0070      0.0083      0.0526
year.2020.0       0.0171     0.0120     1.4197     0.1559     -0.0065      0.0407
year.2021.0      -0.0029     0.0124    -0.2362     0.8133     -0.0272      0.0214
year.2022.0      -0.0030     0.0128    -0.2355     0.8138     -0.0282      0.0221
=================================================================================

F-test for Poolability: 25.991
P-value: 0.0000
Distribution: F(215,1499)

Included effects: Entity
"""

***Marginal F-test for time-FE for eneity cluster on***

F-stat: **38.706**

F-stat (robust): **13.731**

### Marginal test for time FE

In [19]:
exog_vars_marginal_time = ['SIZE', 'PROFITABILITY', 'TANG', 'LIQUID', 'MCAP', 'SOLV', 'compcode']
exog_marginal_time = sm.add_constant(sgx[exog_vars_marginal_time])
fixedeffect(exog = exog_marginal_time, entity_eff = False).summary

<class 'linearmodels.compat.statsmodels.Summary'>
"""
                          PanelOLS Estimation Summary                           
================================================================================
Dep. Variable:               LEVERAGE   R-squared:                        0.8650
Estimator:                   PanelOLS   R-squared (Between):              1.0000
No. Observations:                1728   R-squared (Within):               0.2397
Date:                Mon, Apr 01 2024   R-squared (Overall):              0.8636
Time:                        00:57:40   Log-likelihood                    1628.0
Cov. Estimator:             Clustered                                           
                                        F-statistic:                      43.449
Entities:                         216   P-value                           0.0000
Avg Obs:                       8.0000   Distribution:                F(221,1499)
Min Obs:                       8.0000                                           
Max Obs:                       8.0000   F-statistic (robust):          -1.21e+16
                                        P-value                           1.0000
Time periods:                       8   Distribution:                F(221,1499)
Avg Obs:                       216.00                                           
Min Obs:                       216.00                                           
Max Obs:                       216.00                                           
                                                                                
                               Parameter Estimates                               
=================================================================================
               Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
---------------------------------------------------------------------------------
const            -1.7630     0.4705    -3.7467     0.0002     -2.6860     -0.8400
SIZE              0.0816     0.0259     3.1492     0.0017      0.0308      0.1325
PROFITABILITY    -0.2288     0.0562    -4.0735     0.0000     -0.3390     -0.1186
TANG              0.4097     0.0849     4.8235     0.0000      0.2431      0.5763
LIQUID           -0.0228     0.0066    -3.4833     0.0005     -0.0357     -0.0100
MCAP          -3.265e-12  3.564e-12    -0.9161     0.3598  -1.026e-11   3.727e-12
SOLV             -0.0002     0.0001    -1.5528     0.1207     -0.0004   4.813e-05
compcode.10       0.4267     0.0165     25.925     0.0000      0.3944      0.4590
compcode.12       0.6185     0.0391     15.809     0.0000      0.5418      0.6953
compcode.14       0.3194     0.0466     6.8538     0.0000      0.2280      0.4108
compcode.19      -0.0147     0.0585    -0.2522     0.8009     -0.1294      0.0999
compcode.27       0.4703     0.0550     8.5567     0.0000      0.3625      0.5781
compcode.28       0.6310     0.0320     19.739     0.0000      0.5683      0.6937
compcode.29       0.3182     0.0275     11.571     0.0000      0.2643      0.3722
compcode.33       0.4563     0.0128     35.707     0.0000      0.4312      0.4814
compcode.34       0.4371     0.0554     7.8855     0.0000      0.3284      0.5459
compcode.47       0.4762     0.0854     5.5761     0.0000      0.3087      0.6438
compcode.49       0.3215     0.0508     6.3325     0.0000      0.2219      0.4210
compcode.50       0.3799     0.0676     5.6230     0.0000      0.2474      0.5124
compcode.52       0.6860     0.0729     9.4043     0.0000      0.5429      0.8290
compcode.54       0.0952     0.0131     7.2686     0.0000      0.0695      0.1209
compcode.57       0.5065     0.0295     17.191     0.0000      0.4487      0.5643
compcode.64       0.2215     0.0124     17.848     0.0000      0.1971      0.2458
compcode.65       0.2178     0.0742     2.9355     0.0034      0.0723      0.3634
compcode.67       0.3595     0.0303     11.850     0.0000      0.3000      0.4190
compcode.68       0.3468     

F-statistic = **43.449** for marginal entity effect.